src/theory/quantifiers/sygus/cegis_unif.cpp

```diff
@@ -0,0 +1,252 @@
+                             << std::endl;
+    unsigned sz = d_tds->getSygusTermSize(enum_values[i]);
+    if (i == 0 || sz < min_term_size)
+    {
+      enum_consider.clear();
+      min_term_size = sz;
+      enum_consider.push_back(i);
+    }
+    else if (sz == min_term_size)
+    {
+      enum_consider.push_back(i);
+    }
+  }
+  /* only consider the enumerators that are at minimum size (for fairness) */
+  Trace("cegis-unif-enum") << "...register " << enum_consider.size() << " / "
+                           << enums.size() << std::endl;
+  NodeManager* nm = NodeManager::currentNM();
+  for (unsigned i = 0, ecsize = enum_consider.size(); i < ecsize; ++i)
+  {
+    unsigned j = enum_consider[i];
+    Node e = enums[j];
+    Node v = enum_values[j];
+    std::vector<Node> enum_lems;
+    d_sygus_unif.notifyEnumeration(e, v, enum_lems);
+    /* the lemmas must be guarded by the active guard of the enumerator */
+    Assert(d_enum_to_active_guard.find(e) != d_enum_to_active_guard.end());
+    Node g = d_enum_to_active_guard[e];
+    for (unsigned j = 0, size = enum_lems.size(); j < size; ++j)
+    {
+      enum_lems[j] = nm->mkNode(OR, g.negate(), enum_lems[j]);
+    }
+    lems.insert(lems.end(), enum_lems.begin(), enum_lems.end());
+  }
+  /* build candidate solution */
+  Assert(candidates.size() == 1);
+  Node vc = d_sygus_unif.constructSolution();
+  if (vc.isNull())
+  {
+    return false;
+  }
+  candidate_values.push_back(vc);
+  return true;
+}
+
+Node CegisUnif::purifyLemma(Node n,
+                            bool ensureConst,
+                            std::vector<Node>& model_guards,
+                            BoolNodePairMap& cache)
+{
+  unsigned i, size = n.getNumChildren();
+  Kind k = n.getKind();
+  bool fapp = k == APPLY_UF && size > 0 && n[0] == d_candidate;
+  Trace("cegis-unif-purify") << "PurifyLemma : " << n << "\n";
+  /* top level applications of function-to-synthsize should never be removed */
+  if (ensureConst || !fapp)
+  {
+    BoolNodePairMap::const_iterator it =
+        cache.find(BoolNodePair(ensureConst, n));
+    if (it != cache.end())
+    {
+      Trace("cegis-unif-purify") << "... already visited " << n << "\n";
+      return it->second;
+    }
+  }
+  /* Recurse */
+  bool childChanged = false;
+  std::vector<Node> children;
+  for (i = 0; i < size; ++i)
+  {
+    Node child = purifyLemma(n[i], ensureConst || fapp, model_guards, cache);
+    children.push_back(child);
+    childChanged = childChanged || child != n[i];
+  }
+  Node nb;
+  if (childChanged)
+  {
+    if (n.hasOperator())
+    {
+      children.insert(children.begin(), n.getOperator());
+    }
+    nb = NodeManager::currentNM()->mkNode(k, children);
+    Trace("cegis-unif-purify") << "PurifyFuncApp : transformed " << n
+                               << " into " << nb << "\n";
+  }
+  else
+  {
+    nb = n;
+  }
+  /* get model value if d_candidate applied over constants */
+  if (ensureConst && fapp)
```

Comment:
I think this should just be ensureConst.  Try running it on f( 1+f(0) ).  Here, when we recurse to f(0), we should return the model value of f(0), which is a child of +, which isn't a fapp.

Reply:
This is what happens already. Without the ``fapp`` guard we would, besides replacing ``1+f(0)`` by ``1+[f(0)]^M", also replace ``1+[f(0)]^M`` by ``[1+[f(0)]^M]^M``, which we do not need to do, since the rewriter will do it and this would generate a bogus model equality.

Reply:
Ah, right.  I think once we have the assertion Assert( !ensureConst || nb.isConst() ); at the bottom this will be clear.